sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc

```diff
@@ -240,52 +373,44 @@ int main() {
-  It seems like there are two aspects to the problem.
-  Applications need some way to query which versions the backend (or device)
-  supports.
-  Applications also need some way to tell the runtime which version the kernel
-  is written in.
-+
---
-One option is to define separate enumerators in `source_language` for each
-version like this:
+This example demonstrates how to query the version of OpenCL C that is
+supported, how to query the supported features, and how to query the
+supported extensions.
 
 ```
-enum class source_language : /*unspecified*/ {
-  opencl_1_0,
-  opencl_1_1,
-  opencl_2_0,
-  opencl_3_0,
-};
+#include <iostream>
+#include <sycl/sycl.hpp>
+namespace syclex = sycl::ext::oneapi::experimental;
+
+int main() {
+  sycl::queue q;
+  sycl::device q.get_device();
+
+  if (d.ext_oneapi_can_compile(syclex::source_language::opencl) {
+    std::cout << "Device supports online compilation of OpenCL C kernels\n";
+  }
+  if (d.ext_oneapi_supports_cl_c_version(syclex::opencl_c_3_0) {
+    std::cout << "Device supports online compilation with OpenCL C 3.0\n";
+  }
+  if (d.ext_oneapi_supports_cl_c_feature("__opencl_c_fp64") {
+    std::cout << "Device supports online compilation with 64-bit FP in OpenCL C\n";
+  }
+  syclex::cl_version version;
+  if (d.ext_oneapi_supports_cl_extension("cl_intel_bfloat16_conversions", &version)) {
+    std::cout << "Device supports online compilation of OpenCL C with bfloat16 "
+      "conversions (version: " << version.major << "." << version.minor << "." <<
+      version.patch << ")\n";
+  }
+  if (d.ext_oneapi_cl_profile() == "FULL_PROFILE") {
+    std::cout << "Device supports online compilation with the OpenCL full profile\n";
+  }
+}
 ```
 
-Applications could then query the supported versions via
-`is_source_kernel_bundle_supported`, and applications would identify the
-version of their kernel string via the `lang` parameter to
-`create_kernel_bundle_from_source`.
-
-Alternatively, this extension could define just a single language enumerator
-(`opencl`), but also provide as separate query to get the supported OpenCL C
-versions.
-When building a kernel bundle, applications would be required to pass "-cl-std"
-via the `build_options` property in order to identify the OpenCL C version of
-their source string.
---
-
-* How can an application determine the OpenCL C optional features that are
-  supported and the extensions that are supported?
-  One option is to require the application to use OpenCL APIs for these
-  queries.
-  This seems better than duplicating these queries into this extension.
-  However, this assumes the application is running with an OpenCL backend.
-  Do we want to support the use of OpenCL C kernels also with the Level Zero
-  backend?
-  Currently, the online_compiler does support this case (but it provides no way
-  to query about optional features or extensions).
+
+== Issues
 
 * Do we need to document some restrictions on the OpenCL C
```

Comment:
If I understand it correctly then in the case of using the OpenCL backend there is an option to use interop API for creating a SYCL kernel from an OpenCL one. In that case, I'd expect the backend specification to define how such SYCL kernel can be launched. But given that this extension can be used with the Level Zero backend it seems that it would make sense to define (borrow from the OpenCL backend specification) the behavior for the extension as well.

Reply:
Yes, this is a good observation.  Ideally, we'd improve the OpenCL backend interop specification to also contain information about passing kernel arguments and limitations on using the OpenCL C work-item functions.

Since the "sycl_ext_oneapi_kernel_compiler_opencl" extension seems higher priority, though, I think it might be better to clarify these things first in this extension.  We can improve the OpenCL backend interop spec later, borrowing wording from this extension.  Once that happens, we can change this extension to reference the wording in the OpenCL backend spec, so the wording isn't duplicated.

Reply:
Sounds good.